### PR TITLE
Markings

### PR DESCRIPTION
**Pull request by vulppine**
## About the PR 

Adds species markings to the game. This also adds RGB sliders for skin tones for when a species can use all colors, rather than the colors restricted to human skin tones.

This is a direct port of [fox-station-14](https://github.com/vulppine/fox-station-14)'s marking system, as requested, with modifications to support new features since the last time it was updated, as well as improvements from the last time it was worked on.

### Using markings in species

To use a marking in a species/humanoid prototype, just add the 'Markings' component to an entity's prototype. You can restrict markings as follows:

```yaml
- type: Markings
  layerPoints:
    MarkingCategory:
      points: point count
      required: true/false
      defaultMarkings: [list, of, default, marking, ids]
```

### Creating Markings

Markings are *multi-part* by design, meaning that a single marking can have several groups of color depending on how they're constructed. For example, if you were to add chest hair with two colors:

```yaml
type: marking
id: TwoToneChestHair
bodyPart: Chest
markingCategory: Chest
sprites:
- sprite: Markings/Customization/chest_hair.rsi
  state: hair_l
- sprite: Markings/Customization/chest_hair.rsi
  state: hair_r
```

This would cause two color sliders to appear upon selection, which allows you to modify the color of each individual part. To disable this behavior, however, add the following:

```yaml
followSkinColor: true
```

so that it follows the character's skin color instead.

### Species restrictions

Equally, markings can be restricted to certain species. Simply add a list of species to whitelist like so:

```yaml
speciesRestriction: [Human]
```

#### Naming
Marking names are localization strings: using the example above, you would need to have these localization strings:

```
marking-TwoToneChestHair = Two Tone Chest Hair
marking-TwoToneChestHair-hair_l = Chest Hair Left
marking-TwoToneChestHair-hair_r = Chest Hair Right
```
#### Sprite Order
Markings, when setting up in-game, take priority from top to bottom: this means markings higher up visually will appear on a higher layer on your character.

When constructing them via the prototype, however, it is bottom to top: sprites that are lower will appear below sprites that are higher in the list of sprites.

## TODO:
- [ ] Composite sprite display(?: This doesn't seem to necessarily be supported by ItemList, but I might be wrong)
- [x] Move species specific body parts (i.e., lizard frills) into markings
- [x] Fix category names for `HumanoidVisualLayers`, consolidate for marking placement instead
- [x] Fix categories for markings, instead have appropriate marking sets for mainline SS14

## Screenshots

image

https://user-images.githubusercontent.com/76629141/157817249-ec3a8f9a-ae47-4afb-8265-5b57eaa2c329.mp4

## Changelog

:cl:
- add: Certain crewmembers may find themselves afflicted with mysterious patterns or extrusions on their bodies. Nanotrasen is not responsible for this incident. (You can now add body markings onto your character.)

### CODE REVIEW

**Comment by Pangogie** [comment]
In my opinion the skin colour selectors should be in HSV, which is far more intuitive for finding a specific colour.
And perhaps lizard skin colours should be limited to something more realistic. Cyan - yellow - green - brown - gray.

Just ideas. Good work. :)

**Comment by vulppine** [comment]
> In my opinion the skin colour selectors should be in HSV, which is far more intuitive for finding a specific colour. And perhaps lizard skin colours should be limited to something more realistic. Cyan - yellow - green - brown - gray.
> 
> Just ideas. Good work. :)

thanks! yeah, the RGB fill in is there, as currently, species can either be 'human toned' or 'hues' - HSV might be nice to have, but the RGB slider pattern is all around the character editor at the moment, so i just followed that

realistic colors for other species is another discussion entirely, however

**Comment by moonheart08** [comment]
can we bring lizard back before this gets merged.

**Comment by vulppine** [comment]
> can we bring lizard back before this gets merged.

can't you bring it back with this PR? (also this is still buggy, but i'm trying to work through it now)

**Comment by moonheart08** [comment]
> > can we bring lizard back before this gets merged.
> 
> can't you bring it back with this PR? (also this is still buggy, but i'm trying to work through it now)

it has some kinks to work out like using the wrong limb type

File: Content.Client/Markings/MarkingPicker.xaml.cs [hunk 1]
```diff
@@ -0,0 +1,468 @@
...
+            if (prototype.FollowSkinColor)
+            {
+                CMarkingColors.Visible = false;
+
+                return;
+            }
+
+            var stateNames = GetMarkingStateNames(prototype);
+            _currentMarkingColors.Clear();
+            CMarkingColors.RemoveAllChildren();
```
Comment (metalgearsloth) [inline Content.Client/Markings/MarkingPicker.xaml.cs:344]: Does this dispose the children?

File: Content.Client/Markings/MarkingsSystem.cs [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+
+        public override void Initialize()
+        {
+            SubscribeLocalEvent<MarkingsComponent, ComponentInit>(OnMarkingsInit);
+            SubscribeLocalEvent<MarkingsComponent, SharedHumanoidAppearanceSystem.ChangedHumanoidAppearanceEvent>(UpdateMarkings);
+        }
+
+        private void OnMarkingsInit(EntityUid uid, MarkingsComponent component, ComponentInit __)
+        {
+        }
```
Comment (metalgearsloth) [inline Content.Client/Markings/MarkingsSystem.cs:28]: ?

File: Content.Shared/Markings/MarkingCategories.cs [hunk 3]
```diff
@@ -0,0 +1,19 @@
+using System;
+using Robust.Shared.Serialization;
+
+namespace Content.Shared.Markings
+{
+    [Serializable, NetSerializable]
+    public enum MarkingCategories
```
Comment (metalgearsloth) [inline Content.Shared/Markings/MarkingCategories.cs:7]:
```suggestion
    public enum MarkingCategories : byte
```

File: Content.Shared/Markings/Marking.cs [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+        new public string ToString()
+        {
+            // reserved character
+            string sanitizedName = this.MarkingId.Replace('@', '_');
+            List<string> colorStringList = new();
+            foreach (Color color in _markingColors)
+                colorStringList.Add(color.ToHex());
+
+            return $"{sanitizedName}@{String.Join(',', colorStringList)}";
+        }
```
Comment (metalgearsloth) [inline Content.Shared/Markings/Marking.cs:85]: sussy

File: Content.Shared/Markings/MarkingManager.cs [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+    {
+        [Dependency] private readonly IPrototypeManager _prototypeManager = default!;
+
+        private readonly List<MarkingPrototype> _index = new();
+        private readonly Dictionary<MarkingCategories, List<MarkingPrototype>> _markingDict = new();
+        private readonly Dictionary<string, MarkingPrototype> _markings = new();
+
+        public void Initialize()
+        {
+            _prototypeManager.PrototypesReloaded += OnPrototypeReload;
```
Comment (metalgearsloth) [inline Content.Shared/Markings/MarkingManager.cs:20]: Needs to be unsubscribed, particularly on the client.

File: Content.Shared/Markings/MarkingPrototype.cs [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+
+        [DataField("markingCategory", required: true)]
+        public MarkingCategories MarkingCategory { get; } = default!;
+
+        [DataField("speciesRestriction")]
+        public List<string> SpeciesRestrictions { get; } = new List<string>();
+
+        [DataField("unrestricted", required: true)]
+        public bool Unrestricted { get; } = default!;
+
```
Comment (metalgearsloth) [inline Content.Shared/Markings/MarkingPrototype.cs:30]: Are these related? e.g. if speciesrestrictions were nullable could you assume it's unrestricted?

A few more comments around would be nice too.

File: Content.Shared/Markings/MarkingsComponent.cs [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+using Content.Shared.CharacterAppearance;
+using Robust.Shared.GameObjects;
+using Robust.Shared.Serialization.Manager.Attributes;
+
+namespace Content.Shared.Markings
+{
+    [RegisterComponent]
+    public sealed class MarkingsComponent : Component
+    {
+        public override string Name => "Markings";
```
Comment (metalgearsloth) [inline Content.Shared/Markings/MarkingsComponent.cs:11]: Unnecessary now.

File: Content.Shared/Markings/MarkingsComponent.cs [hunk 8]
```diff
@@ -0,0 +1,44 @@
+using System.Collections.Generic;
+using Content.Shared.CharacterAppearance;
+using Robust.Shared.GameObjects;
+using Robust.Shared.Serialization.Manager.Attributes;
+
+namespace Content.Shared.Markings
+{
+    [RegisterComponent]
```
Comment (metalgearsloth) [inline Content.Shared/Markings/MarkingsComponent.cs:8]: Should it be friend of markingssystem?

**Review by metalgearsloth — CHANGES_REQUESTED** [review]
Should hair be markings now? Seeing as it shows up even for species that don't have it. Also are there any tattoos on vg or the likes just as a reference for human markings?

Also probably need colour restrictions for hues seeing as I can make myself #000000

Someone else also needs to check the migrations too

File: Content.Client/Markings/MarkingPicker.xaml.cs (hunk 1 above)
Comment (vulppine) [reply Content.Client/Markings/MarkingPicker.xaml.cs:344]: nope, whoops

File: Content.Client/Markings/MarkingsSystem.cs (hunk 2 above)
Comment (vulppine) [reply Content.Client/Markings/MarkingsSystem.cs:28]: 👁️

File: Content.Shared/Markings/MarkingPrototype.cs (hunk 6 above)
Comment (vulppine) [reply Content.Shared/Markings/MarkingPrototype.cs:30]: that's a good point - nullable SpeciesRestriction would be pretty much the same thing as 'this is an unrestricted marking'

File: Content.Shared/Markings/MarkingsComponent.cs (hunk 8 above)
Comment (vulppine) [reply Content.Shared/Markings/MarkingsComponent.cs:8]: yeah, but MarkingsSystem is currently clientside only (should this be in client as a result?)

File: Content.Shared/Markings/Marking.cs (hunk 4 above)
Comment (vulppine) [reply Content.Shared/Markings/Marking.cs:85]: workaround for serialization, but i could just rename this / actually figure out serialization stuff (i.e., ColorSerializer) - currently it's just serializing straight to a JSON list because of the need of order preservation

File: Content.Client/Markings/MarkingPicker.xaml.cs [hunk 9]
```diff
@@ -0,0 +1,468 @@
...
+            
+            Populate();
+            List<Marking> toRemove = PopulateUsed();
+            
+            if (toRemove.Count != 0)
+            {
+                foreach (var i in toRemove)
+                {
+                    _usedMarkingList.Remove(i);
+                }
```
[the remark this answers is not in the document]
Comment (vulppine) [reply Content.Client/Markings/MarkingPicker.xaml.cs:308]: ok, that was me being a bit lazy with variable naming - `i` here is just the marking class ref itself, not the index of it

[the remark this answers is not in the document]
Comment (vulppine) [reply Content.Client/Markings/MarkingPicker.xaml.cs:308]: however, removing it while it's iterating through the list and then decrementing the count variable would probably be a little nicer

**Comment by vulppine** [comment]
> Should hair be markings now? Seeing as it shows up even for species that don't have it. Also are there any tattoos on vg or the likes just as a reference for human markings?
> 
> Also probably need colour restrictions for hues seeing as I can make myself #000000
> 
> Someone else also needs to check the migrations too

hair as markings was brought up at some point in time, i think it's possible but it would have to retain the current UI, because IMO, i think that's more convenient to use for picking through hair over the markings UI

**Comment by vulppine** [comment]
would be nice to have #7441 before this is ever merged

File: Content.Shared/Markings/MarkingManager.cs (hunk 5 above)
Comment (metalgearsloth) [reply Content.Shared/Markings/MarkingManager.cs:20]: Actually because it's not part of the sim it's probably gucci

**Review by metalgearsloth — APPROVED** [review]

**Review by DrSmugleaf — APPROVED** [review]
